### PR TITLE
Refactor ERC20

### PR DESCRIPTION
**Pull request by konnov**
This is a refactoring of the pre-Quint specification of ERC20. The new version:

 - Contains a separate specification of ERC20, which may be reused in other specs.
 - Contains a testing setup with mempool, which may be used for bug finding.

This specification should be a goto specification of how Quint could be used for Solidity. Again, there are a few patterns emerging. Once we understand them, we should extract them in a separate module.

### CODE REVIEW

File: examples/solidity/ERC20/erc20.qnt [hunk 1]
```diff
@@ -1,246 +1,359 @@
...
+    // import erc20(AllAddresses = Set("", "a", "b", "c")) as myToken
+    // val myTokenState = myToken::newErc20("b", 1000)
+    const AllAddresses: Set[Address]
+
+    // An EVM integer is 256 bits.
+    // We are using big integers and check for overflows manually.
+    type Uint = int
+    pure def MAX_UINT = 2^256
+
+    // a state of an ERC20 contract/token
```
Comment (thpani) [inline examples/solidity/ERC20/erc20.qnt:28]:
```suggestion
    // A state of an ERC20 contract/token
```

File: examples/solidity/ERC20/erc20.qnt [hunk 2]
```diff
@@ -1,246 +1,359 @@
...
+     */
+    pure def transfer(state: Erc20State, sender: Address,
+            toAddr: Address, amount: Uint): (bool, Erc20Result) = {
+        // `transfer` always returns true, but we should check Erc20Result.error
+        (true, _transfer(state, sender, toAddr, amount))
+    }
+
+    /**
+     * ERC20: Returns the remaining number of tokens that spender will be allowed to
+     * spend on behalf of owner through transferFrom. This is zero by default.
```
Comment (thpani) [inline examples/solidity/ERC20/erc20.qnt:124]: This "zero by default" is from setting it in `newErc20`? Maybe you can explain it here.

File: examples/solidity/ERC20/erc20.qnt [hunk 3]
```diff
@@ -1,246 +1,359 @@
...
+     *
+     * This value changes when approve or transferFrom are called.
```
Comment (thpani) [inline examples/solidity/ERC20/erc20.qnt:126]:
```suggestion
     * This value may change when approve or transferFrom are called.
```

I think?

File: examples/solidity/ERC20/erc20.qnt [hunk 4]
```diff
@@ -1,246 +1,359 @@
...
-      }
-    }
-    not(badExample)
-
-  /* Make sure that the account balances never go negative. */
-  val NoNegativeBalances = ADDR.forall(a => balanceOf.get(a) >= 0)
+    }
+
+    /**
+     * ERC20: Moves amount tokens from the caller’s account to to.
```
Comment (thpani) [inline examples/solidity/ERC20/erc20.qnt:110]:
```suggestion
     * ERC20: Moves amount tokens from the caller’s account to `toAddress`.
```

Not sure, is this what you mean?

File: examples/solidity/ERC20/erc20.qnt [hunk 5]
```diff
@@ -1,246 +1,359 @@
...
+            (true, returnError(err, state))
+        } else {
+            // the case of sender == spender seems to be allowed
+            val newAllowance = state.allowance.set((sender, spender), amount)
+            (true, returnState(state.with("allowance", newAllowance)))
+        }
+    }
+
+    /**
+     * Moves amount tokens from from to to using the allowance mechanism.
```
Comment (thpani) [inline examples/solidity/ERC20/erc20.qnt:155]: idk what "from from to to" means.

Did you mean from `fromAddr` to `toAddr`?

File: examples/solidity/ERC20/README.md [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+This directory contains a Quint specification [erc20.qnt][]
+of an [ERC20 token][]. It contains two modules:
+
+ - `erc20` is an exemplary specification of IERC20. It is designed to be
+   reusable in other specifications. This module is pure: It does not define
+   a state machine.
+
+ - `mempool` is a Quint specification of a single ERC20 token executed in the
+   mempool. This specification defines a state machine. It demonstrates how
+   to specify and detect mempool-related out-of-order issues such as [this one][].
```
Comment (thpani) [inline examples/solidity/ERC20/README.md:12]: Can you explain what `erc20-from-apalache.qnt` is?

**Review by thpani — APPROVED** [review]
LGTM, just a few minor edits

File: examples/solidity/ERC20/erc20.qnt (hunk 5 above)
Comment (konnov) [reply examples/solidity/ERC20/erc20.qnt:155]: Yeah, that's the ERC20 docs. Will fix it :)

**Pull request merged**